student-projects/lilach-elayza/module-III/class-035-14-12-2023/index.ts

```diff
@@ -0,0 +1,111 @@
+import { readFileSync, writeFileSync } from "fs";
+import { createServer } from "http";
+import express from "express";
+import { urlencoded } from "body-parser";
+import { v4 as uuidv4 } from 'uuid';
+
+type Todo = {
+    id: string,
+    name: string,
+    completed: boolean
+}
+
+const todosFilePath = "./todos.json";
+
+const todos = JSON.parse(readFileSync(todosFilePath, "utf-8")) as Todo[];
+
+const app = express();
+
+let visits = 0;
+
+app.set("view engine", "ejs");
+
+const logRequests = (prefix = ""): express.RequestHandler => (req, _, next) => {
+    console.log(prefix, req.method, req.url);
+    next();
+};
+
+app.use(urlencoded());
+
+app.use(logRequests("Todos App"));
+
+app.get("/", (_, res) => {
+    res.render("todos", { todos, visits: ++visits });
+});
+
+
+
+app.post("/addTodo", (req, res) => {
+    const newTodoName = req.body.todo?.toString();
+
+    if (!newTodoName) {
+        res.status(400);
+        res.send("Must provide a todo to add");
+        return;
+    }
+
+    const newTodo = {
+        id: uuidv4(),
+        name: newTodoName, 
+        completed: false
+    }
+
+    todos.push(newTodo);
+    writeFileSync(todosFilePath, JSON.stringify(todos, null ,4));
+
+    res.redirect("/");
+});
+
+app.post("/toggleTodo/:id", (req, res) => {
+    const todoId = req.params.id;
+
+    todos.find(todo => {
```

Comment:
Don't mix manipulation of array with the `find` method.
It could be confusing, but also in this case it prevents you from properly handling the case where a todo was not found.